e2e: reject unknown providers

pohly opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

This finishes the work started for 1.13: instead of merely warning
about an unknown value given to --profile, the test/e2e/e2e.test
binary will now print an error and refuse to run.

That was the intended behavior, we just couldn't do it earlier because it broke some users of the binary. At least kubeadm testing is now fixed (https://github.com/kubernetes/test-infra/pull/10913).

**Which issue(s) this PR fixes**:
<!-- 
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #70200

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--  
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
e2e.test now rejects unknown --provider values instead of merely warning about them. An empty provider name is not accepted anymore and was replaced by "skeleton" (= a provider with no special behavior).
```

/sig testing
/cc @neolit123 @timothysc

pohly committed:
e2e: reject unknown providers

This finishes the work started for 1.13: instead of merely warning
about an unknown value given to --profile, the test/e2e/e2e.test
binary will now print an error and refuse to run.

Fixes: #70200

neolit123 reviewed (commented):
thanks @pohly 

/area test
/priority important-longterm
/lgtm

@kubernetes/sig-testing-pr-reviews 
sent a ping in the #sig-testing chat as well.

neolit123 commented:
`pull-kubernetes-node-e2e` this one is failing on all PRs ATM.

pohly commented:
/test ci-kubernetes-e2e-kubeadm-gce-1-13

Let's see whether the kubeadm job still works. I hope this runs it...

neolit123 commented:
> /test ci-kubernetes-e2e-kubeadm-gce-1-13
> Let's see whether the kubeadm job still works. I hope this runs it...

that job moved to `skeleton`
one that i forgot to modify is:
https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes/sig-testing/bazel-build-test.yaml#L63

sending a PR in a bit.

neolit123 commented:
this needs a minor fix:
https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/73402/pull-kubernetes-verify/120272

pohly commented:
/test pull-kubernetes-node-e2e

neolit123 commented:
> pull-kubernetes-node-e2e

this one might fail for a while due to a GCE zone outage.

timothysc commented on test/e2e/framework/test_context.go:
```diff
@@ -264,7 +266,7 @@ func RegisterClusterFlags() {
 	flag.StringVar(&TestContext.CertDir, "cert-dir", "", "Path to the directory containing the certs. Default is empty, which doesn't use certs.")
 	flag.StringVar(&TestContext.RepoRoot, "repo-root", "../../", "Root directory of kubernetes repository, for finding test files.")
-	flag.StringVar(&TestContext.Provider, "provider", "", "The name of the Kubernetes provider (gce, gke, local, etc.)")
+	flag.StringVar(&TestContext.Provider, "provider", "", "The name of the Kubernetes provider (gce, gke, local, skeleton, etc.)")
```
Why not default to skeleton here?

neolit123 replied:
so both `local` and `skeleton` are null providers here:
https://github.com/kubernetes/kubernetes/blob/b4384e9f828b97c380b93483bea279d2a71d1d00/test/e2e/framework/provider.go#L60-L66

if the previous fallback from `""` as provider was `local` then today setting to `skeleton` as default should not have a negative effect (famous last words).

`local` and `skeleton` in test-infra/kubetest on the other hand *are* different.

pohly replied:
The question is whether anyone cares about having this log message when `--provider` is not set:
https://github.com/kubernetes/kubernetes/blob/b4384e9f828b97c380b93483bea279d2a71d1d00/test/e2e/framework/provider.go#L69

timothysc replied:
If we hard fail we should have a sane default imo.

BenTheElder replied:
local is different here too, because of log dumping in the framework. skeleton is a sane default.

neolit123 replied:
so passing `local` here also does extra steps - e.g. logging? i didn't know that...
my thinking was that it only affects logs on the kubetest side.

BenTheElder replied:
e2e.test will invoke the cluster logs dump script itself, which is """provider""" string aware, including local.

https://github.com/kubernetes/kubernetes/blob/5e86fa43f59f8ed4fd353ab17dc1e65708a9ae71/test/e2e/e2e.go#L154-L162

https://github.com/kubernetes/kubernetes/blob/5e86fa43f59f8ed4fd353ab17dc1e65708a9ae71/test/e2e/framework/util.go#L4760-L4768

neolit123 replied:
i'm convinced that `log-dump.sh` is evil at this point.

neolit123 replied:
@pohly so `""` already does not fallback to "local" if i'm not wrong, as per:
https://github.com/kubernetes/kubernetes/blob/b4384e9f828b97c380b93483bea279d2a71d1d00/test/e2e/framework/provider.go#L68-L71

it effectively falls back to a unnamed null-provider in 1.13 and master, so from what i'm seeing we already modified the old behavior.

@timothysc 
> If we hard fail we should have a sane default imo.

from my understanding we are not failing but instead we are only showing a warning if `--provider` is not set (equals `""`).

pohly replied:
@neolit123 right, it currently basically falls back to "`skeleton` plus warning". Forget what I said about falling back to "local". I think that the current behavior is consistent with <1.13 but haven't checked.

I'm in favor of removing this odd log message and just making `skeleton` the explicit default. Any objections?

neolit123 replied:
+1

pohly replied:
Done. See latest commit in this PR.

timothysc commented on test/e2e/framework/test_context.go:
```diff
@@ -401,20 +403,22 @@ func AfterReadingAllFlags(t *TestContextType) {
-		Failf("Failed to setup fallback skeleton provider config: %v", err)
+		if os.IsNotExist(errors.Cause(err)) {
+			// Provide a more helpful error message when the provider is unknown.
+			var providers []string
```
Please add a //TODO: when providers are extracted this code should be removed.

pohly replied:
Okay.

pohly replied:
I've added a TODO and link to https://github.com/kubernetes/kubernetes/issues/70194

I've kept the wording a bit more open-ended because I can imagine scenarios where kubernetes/kubernetes itself has no provider-specific code, but still supports building custom testsuites where such code is included. That would allow different cloud providers to share common tests in a neutral location when those tests depend on cloud-provider specific code. Whether that is something that Kubernetes as an organization wants to support I don't know.

timothysc commented on test/e2e/framework/test_context.go:
```diff
@@ -401,20 +403,22 @@ func AfterReadingAllFlags(t *TestContextType) {
+		} else {
+			klog.Errorf("Failed to setup provider config for %q: %v", TestContext.Provider, err)
+		}
+		os.Exit(1)
```
I'm not really a fan, but if your long term goal is to extract provider code then I'll be ok with this if there are tracking/umbrealla issues that outline path forwards.

pohly replied:
I personally don't have plans in that direction, but it is getting tracked here: https://github.com/kubernetes/kubernetes/issues/70194

pohly committed:
e2e: change default for --provider

The empty string was the default and then triggered a special
warning. There's no good reason for that behavior, so now the special
handling for "unset provider" is gone and "skeleton" is the non-empty
default for the value.

pohly commented:
/retest

BenTheElder commented:
/retest

pohly commented:
/retest

pohly commented:
/test pull-kubernetes-node-e2e

neolit123 commented:
/skip

neolit123 commented:
@pohly 
it seems the node-e2e test is failing with:

```
I0131 07:24:17.735] [8] E0131 07:24:13.285727    1314 test_context.go:420] Unknown provider "". The following providers are known: local skeleton
I0131 07:24:17.735] 
I0131 07:24:17.735] Ginkgo ran 1 suite in 3.021294514s
I0131 07:24:17.735] Test Suite Failed
```

hm, i though "" should be defaulting to skeleton at this point.

pohly commented:
@neolit123 thanks for investigating the error, I wasn't sure whether it was just flakiness or caused by the PR.

The `--provider` parameter now defaults to `skeleton` when not set. But if it is passed, a valid provider must be specified, so `--provider=` is currently not okay.

Should `--provider=` be allowed? Do we then want to proceed with an empty `TestContext.Provider` string or silently continue as if `--provider=skeleton` had been used?

neolit123 commented:
> Should --provider= be allowed?

pohly committed:
e2e: "skeleton" as fallback, empty string as default

Not accepting --provider= (i.e. setting an empty provider name) broke
some test jobs. As suggested in
https://github.com/kubernetes/kubernetes/pull/73402#issuecomment-459368230,
now --provider= and not passing --provider at all both trigger a
message and then continue as if --provider=skeleton had been used.

pohly commented:
Pushed another commit with that change.

neolit123 commented:
thanks you.
if CI passes and unless others have objections for the above, we can squash the commits and merge the PR.

pohly commented:
/test pull-kubernetes-verify

neolit123 commented:
/lgtm

pohly commented:
@neolit123 do you still want me to squash? I think it is not necessary, each commit makes sense by itself, and I would prefer to merge it like it is now (given that we have clean test results, a lgtm, etc.).

neolit123 commented:
> I think it is not necessary, each commit makes sense by itself, and I would prefer to merge it like it is now

seems fine to me!
would leave the call on that to @timothysc

pohly commented:
@timothysc please have another look and if you agree then approve the PR.

timothysc reviewed (approved):
/approve

neolit123 commented:
looks like this PR still manged to break our signal even if it wasn't supped to:
https://k8s-testgrid.appspot.com/sig-cluster-lifecycle-all#kubeadm-gce-master

will investigate why.
